### ISSUE TITLE
Qwen 3 model not working

### ISSUE THREAD

**Issue by Kenshiro-28** [abetlen/llama-cpp-python#2008]
The new Qwen 3 model is not working with current llama-cpp-python version:

https://qwenlm.github.io/blog/qwen3/

https://huggingface.co/unsloth/Qwen3-235B-A22B-GGUF

Error message:

`
llama_model_load: error loading model: error loading model architecture: unknown model architecture: 'qwen3moe'
`

**Comment by mix1009** [comment]
It does NOT work with llama-cpp-python==0.3.8.

I tried with:
https://huggingface.co/unsloth/Qwen3-4B-GGUF
https://huggingface.co/unsloth/Qwen3-8B-GGUF


It does work with llama.cpp version: 5220 (cdf76586) using llama-server.

**Comment by alonsosilvaallende** [comment]
@mix1009 I'm using a logits processor. It doesn't look like llama-server from llama.cpp server has that option?

**Comment by Kenshiro-28** [comment]
As a workaround you can try to manually build llama-cpp-python with the latest llama.cpp. I tested it only with one prompt but it worked.

```
git clone --recursive https://github.com/abetlen/llama-cpp-python.git
cd llama-cpp-python
git submodule update --remote vendor/llama.cpp
export FORCE_CMAKE=1
export CMAKE_ARGS="-DGGML_BLAS=ON -DGGML_BLAS_VENDOR=OpenBLAS -DLLAMA_CURL=OFF" (first 2 vars if you use OpenBlas)
pip install . --upgrade --force-reinstall --no-cache-dir
```

I hope this helps until official release.

**Comment by alonsosilvaallende** [comment]
> As a workaround you can try to manually build llama-cpp-python with the latest llama.cpp. I tested it only with one prompt but it worked.
> 
> ```
> git clone --recursive https://github.com/abetlen/llama-cpp-python.git
> cd llama-cpp-python
> git submodule update --remote vendor/llama.cpp
> export FORCE_CMAKE=1
> export CMAKE_ARGS="-DGGML_BLAS=ON -DGGML_BLAS_VENDOR=OpenBLAS -DLLAMA_CURL=OFF" (first 2 vars if you use OpenBlas)
> pip install . --upgrade --force-reinstall --no-cache-dir
> ```
> 
> I hope this helps until official release.

@Kenshiro-28 This is gold! Thank you very much!

**Comment by dee-walia20** [comment]
> As a workaround you can try to manually build llama-cpp-python with the latest llama.cpp. I tested it only with one prompt but it worked.
> 
> ```
> git clone --recursive https://github.com/abetlen/llama-cpp-python.git
> cd llama-cpp-python
> git submodule update --remote vendor/llama.cpp
> export FORCE_CMAKE=1
> export CMAKE_ARGS="-DGGML_BLAS=ON -DGGML_BLAS_VENDOR=OpenBLAS -DLLAMA_CURL=OFF" (first 2 vars if you use OpenBlas)
> pip install . --upgrade --force-reinstall --no-cache-dir
> ```
> 
> I hope this helps until official release.

@Kenshiro-28 I tried these steps but I am getting the following error. something to do with CMAKE.

`PS C:\Users\deepak.walia\llama-cpp-python> pip install -e . --no-cache-dir
Obtaining file:///C:/Users/deepak.walia/llama-cpp-python
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: typing-extensions>=4.5.0 in c:\users\deepak.walia\appdata\local\programs\python\python310\lib\site-packages (from llama_cpp_python==0.3.8) (4.13.2)
Requirement already satisfied: numpy>=1.20.0 in c:\users\deepak.walia\appdata\local\programs\python\python310\lib\site-packages (from llama_cpp_python==0.3.8) (2.2.5)
Requirement already satisfied: diskcache>=5.6.1 in c:\users\deepak.walia\appdata\local\programs\python\python310\lib\site-packages (from llama_cpp_python==0.3.8) (5.6.3)
Requirement already satisfied: jinja2>=2.11.3 in c:\users\deepak.walia\appdata\local\programs\python\python310\lib\site-packages (from llama_cpp_python==0.3.8) (3.1.6)
Requirement already satisfied: MarkupSafe>=2.0 in c:\users\deepak.walia\appdata\local\programs\python\python310\lib\site-packages (from jinja2>=2.11.3->llama_cpp_python==0.3.8) (2.0.1)
Building wheels for collected packages: llama_cpp_python
  Building editable for llama_cpp_python (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building editable for llama_cpp_python (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [127 lines of output]
      *** scikit-build-core 0.11.2 using CMake 4.0.0 (editable)
      *** Configuring CMake...
      2025-05-05 14:05:39,938 - scikit_build_core - WARNING - Can't find a Python library, got libdir=None, ldlibrary=None, multiarch=None, masd=None
      loading initial cache file C:\Users\DEEPAK~1.WAL\AppData\Local\Temp\tmpp2po32bi\build\CMakeInit.txt
      -- Building for: Visual Studio 17 2022
      -- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
      -- The C compiler identification is MSVC 19.43.34808.0
      -- The CXX compiler identification is MSVC 19.43.34808.0
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.43.34808/bin/Hostx64/x64/cl.exe - skipped
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.43.34808/bin/Hostx64/x64/cl.exe - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Found Git: C:/Users/deepak.walia/AppData/Local/Programs/Git/cmd/git.exe (found version "2.48.1.windows.1")
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
      -- Looking for pthread_create in pthreads
      -- Looking for pthread_create in pthreads - not found
      -- Looking for pthread_create in pthread
      -- Looking for pthread_create in pthread - not found
      -- Found Threads: TRUE
      -- Warning: ccache not found - consider installing it for faster compilation or disable this warning with GGML_CCACHE=OFF
      -- CMAKE_SYSTEM_PROCESSOR: AMD64
      -- CMAKE_GENERATOR_PLATFORM: x64
      -- Including CPU backend
      -- Found OpenMP_C: -openmp (found version "2.0")
      -- Found OpenMP_CXX: -openmp (found version "2.0")
      -- Found OpenMP: TRUE (found version "2.0")
      -- x86 detected
      -- Performing Test HAS_AVX_1
      -- Performing Test HAS_AVX_1 - Success
      -- Performing Test HAS_AVX2_1
      -- Performing Test HAS_AVX2_1 - Success
      -- Performing Test HAS_FMA_1
      -- Performing Test HAS_FMA_1 - Success
      -- Performing Test HAS_AVX512_1
      -- Performing Test HAS_AVX512_1 - Failed
      -- Performing Test HAS_AVX512_2
      -- Performing Test HAS_AVX512_2 - Failed
      -- Adding CPU backend variant ggml-cpu: /arch:AVX2 GGML_AVX2;GGML_FMA;GGML_F16C
      -- Found CUDAToolkit: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.8/include (found version "12.8.93")
      -- CUDA Toolkit found
      -- Using CUDA architectures: native
      -- The CUDA compiler identification is NVIDIA 12.8.93 with host compiler MSVC 19.43.34808.0
      -- Detecting CUDA compiler ABI info
      -- Detecting CUDA compiler ABI info - done
      -- Check for working CUDA compiler: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.8/bin/nvcc.exe - skipped
      -- Detecting CUDA compile features
      -- Detecting CUDA compile features - done
      -- Including CUDA backend
      CMake Warning (dev) at CMakeLists.txt:13 (install):
        Target llama has PUBLIC_HEADER files but no PUBLIC_HEADER DESTINATION.
      Call Stack (most recent call first):
        CMakeLists.txt:97 (llama_cpp_python_install_target)
      This warning is for project developers.  Use -Wno-dev to suppress it.

      CMake Warning (dev) at CMakeLists.txt:21 (install):
        Target llama has PUBLIC_HEADER files but no PUBLIC_HEADER DESTINATION.
      Call Stack (most recent call first):
        CMakeLists.txt:97 (llama_cpp_python_install_target)
      This warning is for project developers.  Use -Wno-dev to suppress it.

      CMake Warning (dev) at CMakeLists.txt:13 (install):
        Target ggml has PUBLIC_HEADER files but no PUBLIC_HEADER DESTINATION.
      Call Stack (most recent call first):
        CMakeLists.txt:98 (llama_cpp_python_install_target)
      This warning is for project developers.  Use -Wno-dev to suppress it.

      CMake Warning (dev) at CMakeLists.txt:21 (install):
        Target ggml has PUBLIC_HEADER files but no PUBLIC_HEADER DESTINATION.
      Call Stack (most recent call first):
        CMakeLists.txt:98 (llama_cpp_python_install_target)
      This warning is for project developers.  Use -Wno-dev to suppress it.

      CMake Error at CMakeLists.txt:146 (add_subdirectory):
        add_subdirectory given source "vendor/llama.cpp/examples/llava" which is
        not an existing directory.


      CMake Error at CMakeLists.txt:147 (set_target_properties):
        set_target_properties Can not find target to add properties to:
        llava_shared


      CMake Error at CMakeLists.txt:150 (set_target_properties):
        set_target_properties Can not find target to add properties to:
        llava_shared


      CMake Error at CMakeLists.txt:166 (target_include_directories):
        Cannot specify include directories for target "llava" which is not built by
        this project.


      CMake Error at CMakeLists.txt:167 (target_include_directories):
        Cannot specify include directories for target "llava" which is not built by
        this project.


      CMake Error at CMakeLists.txt:170 (target_include_directories):
        Cannot specify include directories for target "llava_shared" which is not
        built by this project.


      CMake Error at CMakeLists.txt:171 (target_include_directories):
        Cannot specify include directories for target "llava_shared" which is not
        built by this project.


      CMake Error at CMakeLists.txt:174 (target_include_directories):
        Cannot specify include directories for target "llama-llava-cli" which is
        not built by this project.


      CMake Error at CMakeLists.txt:175 (target_include_directories):
        Cannot specify include directories for target "llama-minicpmv-cli" which is
        not built by this project.


      -- Configuring incomplete, errors occurred!

      *** CMake configuration failed
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building editable for llama_cpp_python
Failed to build llama_cpp_python
ERROR: Failed to build installable wheels for some pyproject.toml based projects (llama_cpp_python)`

**Comment by AssortedFantasy** [comment]
@dee-walia20 

You can see what options to enable/disable in the `CMakeLists.txt`

In particular you should add this:

`-DLLAMA_LAVA=OFF -DLLAMA_CURL=OFF -DLLAVA_BUILD=OFF` to the `CMAKE_ARGS`

**Comment by dee-walia20** [comment]
Thanks @AssortedFantasy, it worked 👏

**Comment by AleefBilal** [comment]
@alonsosilvaallende 
I followed the steps but during model loading, it is loaded on cpu indead of cuda.
Do you know why?

```
model_path = "/mnt/drive/rev9ai/aleef/Project Zoo/scrapper_content/models/Qwen3-4B-Q4_K_M.gguf"
max_context = 10000
model = Llama(
    model_path=model_path,
    n_ctx=max_context,
    n_gpu_layers=-1, 
    verbose=False,
        )
```

**Comment by Kenshiro-28** [comment]
just a fast update, to use CUDA you need to install llama with something like:

'''
export CMAKE_ARGS="-DGGML_CUDA=on -DLLAMA_CURL=OFF"
pip install . --upgrade --force-reinstall --no-cache-dir
'''

**Comment by alonsosilvaallende** [comment]
@AleefBilal I used the command @Kenshiro-28 suggested:

> I have not tested it but to use CUDA you need to install with something like:
> 
> ```
> export CMAKE_ARGS="-DGGML_CUDA=on -DLLAMA_CURL=OFF"
> pip install . --upgrade --force-reinstall --no-cache-dir
> ```

**Comment by wredan** [comment]
This is the complete pipeline I used to run Qwen3-8B-Q4_K_M.gguf with llama-cpp-python and llama.cpp using CUDA:

```bash
# Setup CUDA 12.4 environment
conda install -c nvidia/label/cuda-12.4.0 cuda-toolkit=12.4
conda install -c nvidia/label/cuda-12.4.0 cuda=12.4

# Configure environment variables
export CUDAToolkit_ROOT=$CONDA_PREFIX
export PATH=$CONDA_PREFIX/bin:$PATH
export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH
export FORCE_CMAKE=1
export CMAKE_ARGS="-DGGML_CUDA=on -DLLAMA_CURL=OFF"

# Clone and update repository
git clone --recursive https://github.com/abetlen/llama-cpp-python.git
cd llama-cpp-python
git submodule update --remote vendor/llama.cpp

# Install package
pip install . --upgrade --force-reinstall --no-cache-dir
```

Additionally, I had to disable flash_attn during import:

```python
from llama_cpp import Llama
model = Llama(
    model_path="Qwen3-8B-Q4_K_M.gguf",
    ...
    flash_attn=False  # Disable flash attention
)
```

Hope this helps

**Comment by Z1EMN1AK** [comment]
> This is the complete pipeline I used to run Qwen3-8B-Q4_K_M.gguf with llama-cpp-python and llama.cpp using CUDA:
> 
> # Setup CUDA 12.4 environment
> conda install -c nvidia/label/cuda-12.4.0 cuda-toolkit=12.4
> conda install -c nvidia/label/cuda-12.4.0 cuda=12.4
> 
> # Configure environment variables
> export CUDAToolkit_ROOT=$CONDA_PREFIX
> export PATH=$CONDA_PREFIX/bin:$PATH
> export LD_LIBRARY_PATH=$CONDA_PREFIX/lib:$LD_LIBRARY_PATH
> export FORCE_CMAKE=1
> export CMAKE_ARGS="-DGGML_CUDA=on -DLLAMA_CURL=OFF"
> 
> # Clone and update repository
> git clone --recursive https://github.com/abetlen/llama-cpp-python.git
> cd llama-cpp-python
> git submodule update --remote vendor/llama.cpp
> 
> # Install package
> pip install . --upgrade --force-reinstall --no-cache-dir
> Additionally, I had to disable flash_attn during import:
> 
> from llama_cpp import Llama
> model = Llama(
>     model_path="Qwen3-8B-Q4_K_M.gguf",
>     ...
>     flash_attn=False  # Disable flash attention
> )
> Hope this helps

I did this but got:
llama_model_load: error loading model: error loading model architecture: unknown model architecture: 'qwen3'

**Comment by Kenshiro-28** [comment]
Qwen3 it's working with the latest release:

https://pypi.org/project/llama-cpp-python/

**Comment by joaojhgs** [comment]
Does anyone know how can I get qwen3 native function calling to work with llama-cpp-python bindings? I don't see any specific chat handlers for it (besides a generic qwen that does not support FC), and it does not work propperly with the default chatml-function-calling chat handler, the model does return a tool call within the text but the chat handler does not deal with it.
